infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesPersistentVolumeClaims.java

```diff
@@ -231,10 +206,63 @@ public void onClose(KubernetesClientException cause) {
       }
     } catch (KubernetesClientException e) {
       throw new KubernetesInfrastructureException(e);
-    } finally {
-      if (watch != null) {
-        watch.close();
-      }
     }
   }
+
+  private Watch pvcIsBoundWatcher(
+      CompletableFuture<PersistentVolumeClaim> future,
+      Resource<PersistentVolumeClaim, DoneablePersistentVolumeClaim> pvcResource,
+      String name) {
+    return pvcResource.watch(
+        new Watcher<PersistentVolumeClaim>() {
+          @Override
+          public void eventReceived(Action action, PersistentVolumeClaim pvc) {
+            if (pvc.getStatus().getPhase().equals(PVC_BOUND_PHASE)) {
+              log.debug("pvc '" + pvc.getMetadata().getName() + "' is bound");
+              future.complete(pvc);
+            }
+          }
+
+          @Override
+          public void onClose(KubernetesClientException cause) {
+            if (cause != null) {
```

Comment:
I'm not sure if we should check cause here. Because even if both of listener is closed normally then somebody should complete future.
Note that the future is already completed then completeExceptionally would do nothing.

Reply:
Good catch. I'm now canceling the future when it's not done and cause is null.